compiler-core/src/recipe.rs

```diff
@@ -0,0 +1,158 @@
+use std::fmt;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use crate::config::PackageConfig;
+use crate::{Error, Result};
+use hexpm::version::Range;
+use serde::de::{self, Deserializer, MapAccess, Visitor};
+use serde::ser::{Serialize, SerializeMap, Serializer};
+use serde::Deserialize;
+
+#[derive(Deserialize, Debug, PartialEq, Eq, Clone)]
+#[serde(untagged, remote = "Self")]
+pub enum Recipe {
+    Hex { version: Range },
+    Path { path: PathBuf },
+    Git { git: String },
+}
+
+impl Recipe {
+    pub fn version_range(&self) -> Result<Range, Error> {
+        match self {
+            Recipe::Hex { version: range } => Ok(range.clone()),
+            Recipe::Path { path } => {
+                let mut config_path = path.clone();
+                config_path.push("gleam.toml");
```

Comment:
```suggestion
                let config_path = path.join("gleam.toml");
```
Bit more concise and remove an allocation.